K2Field.K2NE.ServiceBroker/ServiceObjects/ADOSMOQuerySO.cs

```diff
@@ -1,147 +1,238 @@
+
+            ServiceObject serviceObject = base.ServiceBroker.Service.ServiceObjects[0];
+            string query = serviceObject.Methods[0].MetaData.GetServiceElement<string>("Query");
+            serviceObject.Properties.InitResultTable();
+            DataTable results = base.ServiceBroker.ServicePackage.ResultTable;
+
+            using (SOConnection connection = new SOConnection(base.BaseAPIConnectionString))
+            {
+                using (SOCommand command = new SOCommand(query, connection))
+                {
+                    using (SODataAdapter adapter = new SODataAdapter(command))
+                    {
+                        foreach (Property prop in serviceObject.Properties)
+                        {
+                            if (prop.Value != null)
+                            {
+                                command.Parameters.AddWithValue(prop.Name, prop.Value);
+                            }
+                        }
+                        connection.DirectExecution = true;
+                        connection.Open();
+                        adapter.Fill(results);
+                    }
+                }
+                connection.Close();
+            }
+            return results.CreateDataReader();
+        }
+        private DataTable GetSchema(string query, Dictionary<string, string> props)
+        {
+            DataTable results = new DataTable();
+
+            using (SOConnection connection = new SOConnection(base.BaseAPIConnectionString))
+            {
+                using (SOCommand command = new SOCommand(query, connection))
+                {
+                    using (SODataAdapter adapter = new SODataAdapter(command))
+                    {
+                        foreach (KeyValuePair<string, string> prop in props)
+                        {
+                            if (prop.Value != null)
+                            {
+                                command.Parameters.AddWithValue(prop.Key, prop.Value);
+                            }
+                        }
+
+                        connection.DirectExecution = true;
+                        connection.Open();
+                        adapter.FillSchema(results, SchemaType.Source);
+                    }
+                }
+                connection.Close();
+            }
+            return results;
+        }
+
+        private void ExportToExcel()
+        {
+
+
+            ServiceObject serviceObject = ServiceBroker.Service.ServiceObjects[0];
+            string ADOQuery = serviceObject.Methods[0].MetaData.GetServiceElement<string>("Query");
+            serviceObject.Properties.InitResultTable();
+            DataTable results = ServiceBroker.ServicePackage.ResultTable;
+            DataTable SOQueryResult = new DataTable();
+            string FileName = GetStringProperty(Constants.SOProperties.ExportToExcel.FileName, true);
+
+            using (CreateExcelWrapper excelWrapper = new CreateExcelWrapper())

```

Comment:
The using block code:
We either do all the ADO.NET query stuff INSIDE the helper class or we only pass it the data to generate the excel file. The responsibility of the helper classes and this class is just unclear at the moment. Both have connections with SMO queries and both do something with Excel files.

I'd suggest to change the helper class to a class that has a single public method that allows us to provide a dataset and returns a file stream that we can then (in this class) put into the return properties.